fix(op-challenger): Fetch PreimageOracle Large Preimage Threshold

refcell opened the pull request:
**Description**

This PR updates the `op-challenger` to query the `PreimageOracle` contract for the minimum large preimage size in bytes (the [`MIN_LPP_SIZE_BYTES`](https://github.com/ethereum-optimism/optimism/blob/aa28bcfd1004d55c1d71b1b405c25dabfc12eb8c/packages/contracts-bedrock/src/cannon/PreimageOracle.sol#L22) state variable) which represents a conservative threshold for routing preimage uploads.

The minimum large preimage size is queried during construction of the game player and passed into the `SplitPreimageUploader` to use as its threshold. The `PreimageOracle` and its bindings are also updated in this pr to change the visibility of the `MIN_LPP_SIZE_BYTES` state variable from `internal` to `public`.

**Tests**

Split preimage uploader tests are updated to pass a mock threshold.

**Metadata**

Fixes https://github.com/ethereum-optimism/client-pod/issues/467

refcell committed:
fix(op-challenger): update split preimage uploader to query the preimage oracle contract for the min large preimage size

refcell commented:
Current dependencies on/for this PR:
* `develop`
  * **PR #9204** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/9204?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/9204?utm_source=stack-comment).

refcell commented:
Offline discussion being had.

clabby reviewed (approved):
LGTM

Inphi commented on op-challenger/game/fault/preimages/split.go:
```diff
@@ -8,29 +8,23 @@ import (
 		return ErrNilPreimageData
 	}
-	if len(data.OracleData) > PREIMAGE_SIZE_THRESHOLD {
+	if uint64(len(data.OracleData)) > s.largePreimageSizeThreshold {
```
The PreimageOracle admits large preimages that are exactly `PREIMAGE_SIZE_THRESHOLD`. The challenger doesn't have to use LPPs for that exact size, but perhaps it should to save gas?

Inphi replied:
ref: https://github.com/ethereum-optimism/optimism/blob/1881f2d238b4e2dc2fd0296c2d751ab4e59510b0/packages/contracts-bedrock/src/cannon/PreimageOracle.sol#L271

refcell replied:
It shouldn't make a difference since this is used to provide an overlap buffer between the maximum possible direct preimage upload and the minimum large preimage. So if we compare as greater than with inclusion or not, there must still be a buffer. I will modify it to include the preimage of exact size `s.largePreimageSizeThreshold`, though for sanity.

Inphi commented on packages/contracts-bedrock/src/cannon/PreimageOracle.sol:
```diff
@@ -19,7 +19,7 @@ contract PreimageOracle is IPreimageOracle {
     uint256 internal immutable CHALLENGE_PERIOD;
     /// @notice The minimum size of a preimage that can be proposed in the large preimage path.
-    uint256 internal immutable MIN_LPP_SIZE_BYTES;
+    uint256 public immutable MIN_LPP_SIZE_BYTES;
```
nit: Per style guide, immutables should be internal in favor of explicit getters for them.

refcell replied:
Can do

ajsutton replied:
For the record, we already have `minProposalSize()`

refcell replied:
Oops too late, will revert :)

ajsutton reviewed (commented):
Looks good generally, but something is breaking all the alphabet tests. They always load a local preimage so I'm not quite sure why they'd be affected...

refcell commented:
> Looks good generally, but something is breaking all the alphabet tests. They always load a local preimage so I'm not quite sure why they'd be affected...

Worked through this with @clabby.

So what was happening is the alphabet tests were failing to upload preimages in the `LargePreimageUploader` when they were failing to get proposal metadata. These preimages were `local` which should **only** be routed to the direct preimage uploader sine the FDG needs to load them directly.

The reason why this was previously working before this pr, is the direct preimage uploader has a branch inside the manual contract bindings to craft a local preimage data upload transaction if the data is local.

To fix this, I've modified the `SplitPreimageUploader` to route preimage uploads to the `DirectPreimageUploader` if the preimage data `IsLocal` field is truthy. I've also added a unit test against the `SplitPreimageUploader` to verify this new branch condition works as expected.

Inphi reviewed (approved):
LGTM!

ajsutton commented on op-challenger/game/fault/preimages/split.go:
```diff
@@ -8,31 +8,26 @@ import (
-		return s.largeUploader.UploadPreimage(ctx, parent, data)
-	} else {
+	// Always route local preimage uploads to the direct uploader.
+	if uint64(len(data.OracleData)) < s.largePreimageSizeThreshold || data.IsLocal {
```
I'm fine with merging this as it does make sense to hard code local images as using direct upload. However, given `OracleData` should be `nil` for local data (since we only need to provide the key and the data from from the game contract), why would it ever use large preimages?
